### PR TITLE
Various small fixes to double buffering getRange.

### PR DESCRIPTION
**Pull request by nemanja-boric-sociomantic**
cc @david-eckardt-sociomantic

### CODE REVIEW

File: src/dlsproto/client/request/internal/GetRange.d [hunk 1]
```diff
@@ -633,7 +633,7 @@ private scope class GetRangeHandler
         private bool waitForRecords ()
         {
             // Wait for the next batch, unless we already have one.
-            if ((*this.buffers.input).length == 0)
+            if (this.buffers.empty())
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d:638]: Yes. Oops.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 2]
```diff
@@ -465,6 +465,11 @@ private scope class GetRangeHandler
                 if (!(*input).length)
                     enableStomping(*input);
 
+                // note that appending the data here is not needed, since
+                // if the node has received Stop signal, _it will not send the
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d]: Rogue `_`

File: src/dlsproto/client/request/internal/GetRange.d (hunk 1 above)
Comment (gavin-norman-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d:638]: I guess this fixes the bug revealed in testing?

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d:638]: No, it haven't tested it yet, but I'm not so confident this will fix that bug.

**Comment by nemanja-boric-sociomantic** [comment]
Updated to appending into the background patch (my first remark & comment were wrong, there's a case when we send out the batch after received stopped message: https://github.com/sociomantic-tsunami/dlsproto/blob/neo/src/dlsproto/node/neo/request/GetRange.d#L382-L384

File: src/dlsproto/client/request/internal/GetRange.d [hunk 3]
```diff
@@ -465,8 +465,12 @@ private scope class GetRangeHandler
                 if (!(*input).length)
                     enableStomping(*input);
 
-                (*input).length = data.length;
-                (*input)[] = data[];
+                // Note that the appending here is needed, since the node
+                // will send the remaining batch if received the Stop message
+                // during sending the batch.
+                auto old_len = (*input).length;
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d]: Isn't this the same as doing `(*input) ~= data;`?

File: src/dlsproto/client/request/internal/GetRange.d [hunk 4]
```diff
@@ -465,8 +465,12 @@ private scope class GetRangeHandler
                 if (!(*input).length)
                     enableStomping(*input);
 
-                (*input).length = data.length;
-                (*input)[] = data[];
+                // Note that the appending here is needed, since the node
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d:468]: Maybe change the method name to `append`, or `push`, or something?

File: src/dlsproto/client/request/internal/GetRange.d (hunk 3 above)
Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d]: Yeah, good point :man_facepalming:

**Comment by nemanja-boric-sociomantic** [comment]
Updated.

File: src/dlsproto/client/request/internal/GetRange.d [hunk 5]
```diff
@@ -451,22 +451,24 @@ private scope class GetRangeHandler
...
                 // be empty.
                 if (!(*input).length)
                     enableStomping(*input);
 
-                (*input).length = data.length;
-                (*input)[] = data[];
+                // Note that the appending here is needed, since the node
+                // will send the remaining batch if received the Stop message
+                // during sending the batch.
+                (*input) ~= data[];
```
Comment (gavin-norman-sociomantic) [inline src/dlsproto/client/request/internal/GetRange.d]: Is the `[]` necessary?

Comment (nemanja-boric-sociomantic) [reply src/dlsproto/client/request/internal/GetRange.d]: Removed.

**Review by gavin-norman-sociomantic — APPROVED** [review]

**Comment by gavin-norman-sociomantic** [comment]
Approved. Needs rebase now.

**Comment by nemanja-boric-sociomantic** [comment]
Rebased

**Comment by gavin-norman-sociomantic** [comment]
Feel free to auto-merge.

**Pull request merged**